src/AutoUI/Lenses/types/table.tsx

```diff
@@ -2,13 +2,167 @@ import React from 'react';
-import { Table } from 'rendition';
-import { setToLocalStorage } from '../../utils';
-import pick from 'lodash/pick';
-import { useAnalyticsContext } from '@balena/ui-shared-components';
+import {
+	DEFAULT_ITEMS_PER_PAGE,
+	getFromLocalStorage,
+	setToLocalStorage,
+} from '../../utils';
+import { Table } from '../../../components/Table';
+import type { AutoUIEntityPropertyDefinition } from '~/AutoUI';
+import type {
+	TableSortOptions,
+	TagField,
+} from '../../../components/Table/utils';
+import { useColumns } from '../../../components/Table/useColumns';
+import { useTagActions } from '../../../components/Table/useTagActions';
+import { AddTagHandler } from '../../../components/Table/AddTagHandler';
 
-const TableColumnStateStoredProps = ['key', 'selected', 'type', 'tagKey'];
+import {
+	designTokens,
+	Material,
+	Copy,
+	useAnalyticsContext,
+} from '@balena/ui-shared-components';
+const { Box, styled, Tooltip, Typography } = Material;
+
+const TAG_COLUMN_PREFIX = 'tag_column_';
+
+const TagContainer = styled(Box)(() => ({
+	border: `0.5px solid ${designTokens.color.border.accent.value}`,
+	borderRadius: '3px',
+	color: designTokens.color.text.value,
+	backgroundColor: designTokens.color.bg.accent.value,
+	position: 'relative',
+	width: 'fit-content',
+}));
+
+export interface TagLabelProps {
+	value: string;
+}
+
+export const TagLabel = ({ value }: TagLabelProps) => {
+	return (
+		<Tooltip title={value}>
+			<TagContainer>
+				<Copy copy={value} variant="absolute">
+					<Box display="flex" gap={1} mx={2} my={1}>
+						<Typography
+							variant="bodySm"
+							noWrap
+							fontWeight="bold"
+							maxWidth="100px"
+						>
+							{value}
+						</Typography>
+					</Box>
+				</Copy>
+			</TagContainer>
+		</Tooltip>
+	);
+};
+
+const tagKeyRender = (key: string) => {
+	function renderFunction(
+		tags: Array<{ tag_key: string; value: string }> | undefined,
+	) {
+		if (!tags?.length) {
+			return null;
+		}
+		const tag = tags?.find((t) => t.tag_key === key);
+		return tag ? <TagLabel value={tag.tag_key} /> : null;
+	}
+
+	return renderFunction;
+};
+
+const getResourceTags = <T extends object, P extends keyof T>(
+	item: T,
+	tagField: P,
+) => ((tagField in item ? item[tagField] : null) ?? []) as TagField[];
+
+const findTagOfTaggedResource = <T extends object>(
+	taggedResource: T,
+	tagField: keyof T,
+	tagKey: string,
+) =>
+	getResourceTags(taggedResource, tagField)?.find(
+		(tag) => tag.tag_key === tagKey,
+	);
+
+const sortData = <T extends object>(
+	data: T[],
+	columns: Array<AutoUIEntityPropertyDefinition<T>>,
+	sort: TableSortOptions | null,
+) => {
+	if (sort?.field == null) {
+		return data;
+	}
+
+	const column = columns.find((c) => c.key === sort.key);
+
+	if (!column) {
+		return data;
+	}
+
+	let collection;
+
+	if ('sortable' in column && typeof column.sortable === 'function') {
+		collection = data.slice().sort(column.sortable);
+	} else if (column.key.startsWith(TAG_COLUMN_PREFIX)) {
```

Comment:
We're including this even though we don't have serverside sorting by tag? Sounds like we're enabling single-page clientside sorting by tag? Or will this *only* work for clientside tables?

Reply:
Nevermind, I see we only call this function for clientside sorting

Reply:
we don't run this function on a serverside context

https://github.com/balena-io-modules/autoui/blob/a298359ecf1e09433789197a4155adacc9022385/src/AutoUI/Lenses/types/table.tsx#L198